micro_dl/train/metrics.py

```diff
@@ -58,39 +73,101 @@ def dice_coef(y_true, y_pred, smooth=1.):
     return dice
 
 
-def ssim(y_true, y_pred, max_val=6):
+def flip_dimensions(func):
+    """
+    Decorator to convert channels first tensor to channels last.
+
+    :param func: Function to be decorated
+    """
+    def wrap_function(y_true, y_pred):
+        """
+        Shifts dimensions to channels last if applicable, before
+        calling func.
+
+        :param tensor y_true: Gound truth data
+        :param tensor y_pred: Predicted data
+        :return: function called with channels last
+        """
+        if K.image_data_format() == 'channels_first':
+            if K.ndim(y_true) > 4:
+                y_true = tf.transpose(y_true, [0, 2, 3, 4, 1])
+                y_pred = tf.transpose(y_pred, [0, 2, 3, 4, 1])
+            else:
+                y_true = tf.transpose(y_true, [0, 2, 3, 1])
+                y_pred = tf.transpose(y_pred, [0, 2, 3, 1])
+        return func(y_true, y_pred)
+    return wrap_function
+
+
+def _get_max_min_val(y_true, y_pred):
+    """
+    Get intensity range and min of ground truth and prediction
+
+    :param tensor y_true: Gound truth data
+    :param tensor y_pred: Predicted data
+    :return float max_val: Value range (max - min)
+    :return float min_val: Minimum value of true and pred
+    """
+    min_tensor = tf.stack([K.min(y_true), K.min(y_pred)])
+    max_tensor = tf.stack([K.max(y_true), K.max(y_pred)])
+    min_val = K.min(min_tensor)
+    max_val = K.max(max_tensor) - min_val
+    return max_val, min_val
+
+
+@flip_dimensions
+def ssim(y_true, y_pred):
     """Structural similarity
-    Use max_val=5 to approximate maximum of normalized images
-    tensorflow does not support SSIM for 3D images. Need a different
-    way to compute SSIM for 5D tensor
-    :param tensor y_true: Labeled ground truth
-    :param tensor y_pred: Predicted labels, potentially non-binary
-    :param float max_val: The dynamic range of the images (i.e., the
-        difference between the maximum the and minimum allowed values).
+    Uses a default max_val=6 to approximate maximum of normalized images.
```

Comment:
max_val here is the dynamic range of input image. B'coz we use z-scored images, setting this to 6 covers 3 standard deviations.